src/symbol/cross_tile_symbol_index.js

```diff
@@ -0,0 +1,249 @@
+
+    When we remove a label that's currently showing, we look for duplicates
+    at higher resolution, and if we find one we copy our opacity state
+    to that label.
+
+    The code mostly assumes that at any given time a "unique symbol" will have
+    one "non-duplicate" entry, and that the rest of the entries in the
+    index will all be marked as duplicate. This is not necessarily true:
+
+    (1) The code searches child/parent hierarchies for duplicates, but it
+        is possible for the source to contain symbols with anchors on tile
+        boundaries, where the symbol does not stay in the same hierarchy
+        at all zoom levels.
+    (2) A high resolution tile may contain two symbols with the same label
+        but different anchor points. At lower resolution, both of those
+        symbols will appear to be the same.
+
+    In the cases that violate our assumptions, copying opacities between
+    zoom levels won't work as expected. However, the highest resolution
+    tile should always "win", so that after some fade flicker the right
+    label will show.
+*/
+
+// Round anchor positions to roughly 4 pixel grid
+const roundingFactor = 512 / EXTENT / 2;
+
+class TileLayerIndex {
+    coord: TileCoord;
+    sourceMaxZoom: number;
+    symbolInstances: any;
+
+    constructor(coord: TileCoord, sourceMaxZoom: number, symbolInstances: Array<SymbolInstance>) {
+        this.coord = coord;
+        this.sourceMaxZoom = sourceMaxZoom;
+        this.symbolInstances = {};
+
+        for (const symbolInstance of symbolInstances) {
+            const key = symbolInstance.key;
+            if (!this.symbolInstances[key]) {
+                this.symbolInstances[key] = [];
+            }
+            // This tile may have multiple symbol instances with the same key
+            // Store each one along with its coordinates
+            this.symbolInstances[key].push({
+                instance: symbolInstance,
+                coordinates: this.getScaledCoordinates(symbolInstance, coord, coord.z)
+            });
+            symbolInstance.isDuplicate = false;
+            // If we don't pick up an opacity from our parent or child tiles
+            // Reset so that symbols in cached tiles fade in the same
+            // way as freshly loaded tiles
+            symbolInstance.textOpacityState = new OpacityState();
+            symbolInstance.iconOpacityState = new OpacityState();
+        }
+    }
+
+    getScaledCoordinates(symbolInstance: SymbolInstance, coord: TileCoord, z: number): any {
+        const scale = Math.pow(2, Math.min(this.sourceMaxZoom, z) - Math.min(this.sourceMaxZoom, coord.z)) * roundingFactor;
+        const anchor = symbolInstance.anchor;
+        return {
+            x: Math.floor((coord.x * EXTENT + anchor.x) * scale),
+            y: Math.floor((coord.y * EXTENT + anchor.y) * scale)
+        };
+    }
+
+    get(otherTileSymbol: SymbolInstance, coord: TileCoord) {
```

Comment:
any ideas for a better name for this method?

Reply:
`getMatchingSymbol`?